### ISSUE TITLE
Slow Speed Issue, When download near to complete

### ISSUE THREAD

**Issue by shahwaliullah** [erickutcher/httpdownloader#172]
I get good speed, when i choose 8 segment/connection, it shows 8/8 connection,
when it go to near 95-96% download, speed drop down, for example, if speed was 500 kb/s, then when 95-96% download, completed, speed goes down to 40-50 kb/s, and it shows connection 1/8

**Comment by erickutcher** [comment]
That's normal behavior. The first 7 connections finished downloading their segment and only 1 connection remained. The speed of each connection cumulatively adds up. If there's only one connection going, then it'll only be as fast as that connection.

What you can do to speed it up again is stop the download, and then start it. The remaining data will be split between the number of connections you supplied and you should get faster speeds.

**Comment by JNavas2** [comment]
@erickutcher 
Yes, that can be an effective manual workaround, but it would be better to _automatically reallocate connections as they complete_. Please consider adding that, at least as an option. Thank you.

**Comment by erickutcher** [comment]
I'll have to give it some thought. I can see some potential issues having it automated so I'd need to think of an appropriate way to handle it.

**Comment by JNavas2** [comment]
@erickutcher 
Understood. Thank you!

**Comment by erickutcher** [comment]
A new version is up that implements the reallocation feature. It's "Reallocate parts to maximize connections" under the Connection Options.

**Comment by JNavas2** [comment]
> A new version is up that implements the reallocation feature. It's "Reallocate parts to maximize connections" under the Connection Options.

Wow, you rock! It's almost perfect.
The only problem I see is on completion. It hit 100% on an 8-part download, then stalled, very slowly closing all the remaining connections one by one. Screen shot shortly after it hit 100%:

<img width="584" alt="2021-10-30-1747 063" src="https://user-images.githubusercontent.com/14850792/139562412-446b7f98-61ae-41c1-bba5-75098c19f788.png">
:

**Comment by JNavas2** [comment]
Video of stalling at completion: https://youtu.be/KMRJ7urC97A

**Comment by erickutcher** [comment]
Does it consistently do this? I've tried downloading a bunch of different Linux isos to see if I can reproduce it, but it's not happening for me.

Do you have a URL that I can download to test this?

**Comment by JNavas2** [comment]
> Does it consistently do this? I've tried downloading a bunch of different Linux isos to see if I can reproduce it, but it's not happening for me.

It's consistent on a particular hosting site. I have not tried others.

> Do you have a URL that I can download to test this?

Send me an email (link on my website), and I tell you how to reproduce it.

**Comment by erickutcher** [comment]
I think I may know what's going on. If the site is responding with a "Connection: close" header value, then the program is over-downloading the parts that are split. It all works properly if the site accepts a keep-alive connection...which most do. This won't happen with FTP or SFTP connections though. It's just an HTTP thing.

I'll try to get a new version up in a couple days to fix this.

**Comment by JNavas2** [comment]
> I'll try to get a new version up in a couple days to fix this.
Thank you!

**Comment by JNavas2** [comment]
I just tested an HTTP download from Google Drive, and it completed properly,
so this must be something unique to particular websites.

**Comment by erickutcher** [comment]
Give this beta a try. I found a simple fix for the problem I described and I think it should resolve what you've been experiencing. [HTTP_Downloader_64_1.0.4.7_Beta.zip](https://github.com/erickutcher/httpdownloader/files/7448045/HTTP_Downloader_64_1.0.4.7_Beta.zip)

**Comment by JNavas2** [comment]
Perfect!
I will show my appreciation with a donation.

**Comment by erickutcher** [comment]
Thank you. I do appreciate that. I'm glad it's working. I'll have a new release up some time tomorrow.

**Comment by erickutcher** [comment]
A new version is up. Thanks again for the donation.

**Comment by JNavas2** [comment]
I've run into an apparent edge case where reallocation fails.
With 3 downloads, I stopped all 3 so I could reorder them (absent queuing), then started #2, #3, and #1, in that order.
Some time later I checked back, and found #2 stuck in only 4 out of 8 parts active. See below.
I stopped #1 and #2, then started #2 and #1 in that order, and #2 then completed with 8 active until near the end.
It's not a huge deal, but let me know if I can help in any way.

<img width="587" alt="HTTP_Downloader_20211104-100659 227" src="https://user-images.githubusercontent.com/14850792/140389306-8712910a-1b0a-450a-a521-33d76c615d25.png">

**Comment by erickutcher** [comment]
I'm thinking that maybe the remote server closed those connections before their parts could complete. The reallocation only happens when a part fully downloads. By default the program doesn't retry incomplete parts, but you can change that in the Connection options.

**Comment by JNavas2** [comment]
I have that option on. See below.
Maybe it has something to do with Stopping and then Starting the download?

<img width="545" alt="HTTP_Downloader_20211104-142157 935" src="https://user-images.githubusercontent.com/14850792/140422901-4f467e3b-ab84-4a5a-a100-b69e7525b4a7.png">

**Comment by erickutcher** [comment]
It's most likely a connection issue with the remote server. I can get it to happen if I use a program like CurrPorts to kill some of the connections, but stopping and starting the downloads don't seem to have any effect on my end. It all seems to work as it should. I'd probably just increase the retry count to see if it stabilizes things.

**Comment by JNavas2** [comment]
Thank you. As I said, it's not a huge deal.
What you've already done has made the huge difference for me.
